tools/migration/internal/runner_test.go

```diff
@@ -1,18 +1,178 @@
+		runner := NewMigrationRunner(false, "describe", repoDir)
+		runner.MigrationsProvider = testProviderPasses
+
+		require.NoError(t, ioutil.WriteFile(filepath.Join(repoDir, "version"), []byte("-1"), 0644))
+		assert.EqualError(t, runner.Run(), "repo version out of range: -1")
+
+		require.NoError(t, ioutil.WriteFile(filepath.Join(repoDir, "version"), []byte("32767"), 0644))
+		assert.EqualError(t, runner.Run(), "repo version out of range: 32767")
+	})
+
+	t.Run("Returns error when Migration fails", func(t *testing.T) {
+		require.NoError(t, ioutil.WriteFile(filepath.Join(repoDir, "version"), []byte("0"), 0644))
+		runner := NewMigrationRunner(false, "migrate", repoDir)
+		runner.MigrationsProvider = testProviderMigrationFails
+		assert.EqualError(t, runner.Run(), "migration has failed")
+	})
+
+	t.Run("Returns error when Validation fails", func(t *testing.T) {
+		require.NoError(t, ioutil.WriteFile(filepath.Join(repoDir, "version"), []byte("0"), 0644))
+		runner := NewMigrationRunner(false, "migrate", repoDir)
+		runner.MigrationsProvider = testProviderValidationFails
+		assert.EqualError(t, runner.Run(), "validation has failed")
+	})
+
+	t.Run("Subsequent calls to Migrate migrate subsequent migrations", func(t *testing.T) {
+
+	})
+
+	t.Run("Returns error if version file does not contain an integer string", func(t *testing.T) {
+
+	})
+}
+
+func testProviderPasses() []Migration {
+	return []Migration{
+		&TestMigration01{},
+	}
+}
+
+func testProviderValidationFails() []Migration {
+	return []Migration{&TestMigValidationFails{}}
+}
+
+func testProviderMigrationFails() []Migration {
+	return []Migration{&TestMigMigrationFails{}}
+}
+
+type TestMigration01 struct {
+}
+
+func (m *TestMigration01) Describe() string {
+	return "the migration that doesn't do anything"
+}
+
+func (m *TestMigration01) Migrate(newRepoPath string) error {
+	return nil
+}
+func (m *TestMigration01) Versions() (from, to uint) {
+	return 0, 1
+}
+
+func (m *TestMigration01) Validate(oldRepoPath, newRepoPath string) error {
+	return nil
+}
+
+type TestMigMigrationFails struct {
+}
+
+func (m *TestMigMigrationFails) Versions() (from, to uint) {
+	return 0, 1
+}
+
+func (m *TestMigMigrationFails) Describe() string {
+	return "the migration that doesn't do anything"
+}
+
+func (m *TestMigMigrationFails) Migrate(newRepoPath string) error {
+	return errors.New("migration has failed")
+}
+
+func (m *TestMigMigrationFails) Validate(oldRepoPath, newRepoPath string) error {
+	return nil
+}
+
+type TestMigValidationFails struct {
+}
+
+func (m *TestMigValidationFails) Versions() (from, to uint) {
+	return 0, 1
+}
+
+func (m *TestMigValidationFails) Describe() string {
+	return "the migration that doesn't do anything"
+}
+
+func (m *TestMigValidationFails) Migrate(newRepoPath string) error {
+	return nil
+}
+
+func (m *TestMigValidationFails) Validate(oldRepoPath, newRepoPath string) error {
+	return errors.New("validation has failed")
```

Comment:
You could compress this code by defining the migration at runtime, with a TestMigration struct that has version, description, and function pointers.